cpp/src/parquet/level_conversion.h

```diff
@@ -132,43 +137,49 @@ struct PARQUET_EXPORT LevelInfo {
   }
 };
 
-void PARQUET_EXPORT DefinitionLevelsToBitmap(
-    const int16_t* def_levels, int64_t num_def_levels, const int16_t max_definition_level,
-    const int16_t max_repetition_level, int64_t* values_read, int64_t* null_count,
-    uint8_t* valid_bits, int64_t valid_bits_offset);
-
-// These APIs are likely to be revised as part of ARROW-8494 to reduce duplicate code.
-// They currently represent minimal functionality for vectorized computation of definition
-// levels.
-
-#if defined(ARROW_LITTLE_ENDIAN)
-/// Builds a bitmap by applying predicate to the level vector provided.
-///
-/// \param[in] levels Rep or def level array.
-/// \param[in] num_levels The number of levels to process (must be [0, 64])
-/// \param[in] predicate The predicate to apply (must have the signature `bool
-/// predicate(int16_t)`.
-/// \returns The bitmap using least significant "bit" ordering.
-///
-/// N.B. Correct byte ordering is dependent on little-endian architectures.
-///
-template <typename Predicate>
-uint64_t LevelsToBitmap(const int16_t* levels, int64_t num_levels, Predicate predicate) {
-  // Both clang and GCC can vectorize this automatically with SSE4/AVX2.
-  uint64_t mask = 0;
-  for (int x = 0; x < num_levels; x++) {
-    mask |= static_cast<uint64_t>(predicate(levels[x]) ? 1 : 0) << x;
-  }
-  return mask;
-}
-
-/// Builds a  bitmap where each set bit indicates the corresponding level is greater
-/// than rhs.
-static inline uint64_t GreaterThanBitmap(const int16_t* levels, int64_t num_levels,
-                                         int16_t rhs) {
-  return LevelsToBitmap(levels, num_levels, [rhs](int16_t value) { return value > rhs; });
-}
+/// Input/Output structure for reconstructed validity bitmaps.
+struct PARQUET_EXPORT ValidityBitmapInputOutput {
+  /// The maximum number of values_read expected (actual
+  /// values read must be less than or equal to this value.
+  /// If this number is exceeded methods will throw a
+  /// ParquetException.
+  int64_t values_read_upper_bound = 0;
+  /// The number of values added to the bitmap.
+  int64_t values_read = 0;
+  /// The number of nulls encountered.
+  int64_t null_count = 0;
+  // The validity bitmp to populate. Can only be null
+  // for DefRepLevelsToListInfo (if all that is needed is list lengths).
+  uint8_t* valid_bits = NULLPTR;
+  /// Input only, offset into valid_bits to start at.
+  int64_t valid_bits_offset = 0;
+};
 
+/// Converts def_levels to validity bitmaps for non-list arrays.
+void PARQUET_EXPORT DefinitionLevelsToBitmap(const int16_t* def_levels,
```

Comment:
The variations in terminology confuse me a lot. There's "DefinitionLevelsToBitmap" and "ConvertDefRepLevelsToBitmap". Can we settle on a clear naming scheme, so that it's apparent what the actual differences are?